rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,136 @@
+            fd: -1,
+            locked: false,
+        }
+    }
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.locked {
+            remove_lock(&self.lock_filepath);
+            let _ = unsafe { libc::close(self.fd) };
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock<P: AsRef<Path>>(
+        &mut self,
+        new_lock_filepath: P,
+        pid: &mut libc::pid_t,
+    ) -> bool {
+        // switched to string as it is easier to work with c ffi later on
+        let new_lock_filepath = new_lock_filepath.as_ref().to_str().unwrap();
+        if self.locked && self.lock_filepath == new_lock_filepath {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        // TODO: move to safe rust later
+        // let openoptions = OpenOptions::new().read(true).write(true).create(true);
+        // let mut file = match openoptions.open(path) {
+        //     Ok(file) => file,
+        //     Err(_) => return false,
+        // };
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_filepath
+        );
+
+        // first we open (and possibly create) the lock file
+        self.fd = unsafe {
+            libc::open(
+                new_lock_filepath.as_ptr() as *const i8,
+                libc::O_RDWR | libc::O_CREAT,
+                0600,
+            )
+        };
+        if self.fd < 0 {
+            return false;
+        }
+
+        // then we lock it (returns immediately if locking is not possible)
+        if unsafe { libc::lockf(self.fd, libc::F_TLOCK, 0) } == 0 {
+            log!(
+                Level::Debug,
+                "FsLock: locked `{}', writing PID...",
+                new_lock_filepath
+            );
+            let pidtext = process::id().to_string();
+            let mut written = 0;
+            if unsafe { libc::ftruncate(self.fd, 0) } == 0 {
+                written = unsafe {
+                    libc::write(
+                        self.fd,
+                        pidtext.as_ptr() as *const libc::c_void,
+                        pidtext.len(),
+                    )
+                };
+            }
+            let success = written != -1 && written as usize == pidtext.len();
+            log!(
+                Level::Debug,
+                "FsLock: PID written successfully: {}",
+                success as usize
+            );
+            if success {
+                if self.locked {
+                    remove_lock(&self.lock_filepath);
+                }
+                self.locked = true;
+                self.lock_filepath = new_lock_filepath.to_owned();
+            } else {
+                let _ = unsafe { libc::close(self.fd) };
+            }
+            success
+        } else {
+            log!(Level::Error, "FsLock: something went wrong during locking");
+            // TODO errno
```

Comment:
I removed the TODO.

Reply:
I still can't merge because this *has* to be implemented. Without the errno in the logs, we won't be able to figure out why locking failed.

Please don't mark this comment resolved until it actually is.

Reply:
Do you have any suggestions how do we add this?

Reply:
I think we discussed some options on IRC. Did you try any of them? What results did you get?

Reply:
I think I tried some but I don't remember already.

Reply:
How exactly can this be fixed when we switch to safe Rust?

Is there anything I can do to help you move forward with this? Dig up the IRC logs to see what options we discussed, perhaps?

Reply:
Let me try and see if moving to safe rust works first.

Reply:
No, safe rust does not have this either.

Reply:
I'm not entirely happy with adding three more crates just to read the `errno`. Our IRC discussion mentioned [this post](https://github.com/rust-lang/libc/issues/1644#issue-552020219), which says that `std::io::last_os_error()` provides a read-only access. Did you try that? Does it work?

Also, the original code printed a textual representation of the error into the log, and the new one just prints a number. This should be fixed, if possible.

Reply:
No, haven't checked that. Let me see.